README.md

```diff
@@ -1,111 +1,242 @@
+        decls.NewIdent("auth.claims",
+            decls.NewMapType(decls.String, decls.Dyn), nil)))
+```
+
+That's it, the environment is ready to be use for parsing and type-checking.
+Note, the `auth.claims` variable is exposed as a map with `string` keys and
+`dyn` values. The `dyn` in this case represents JSON input as the
+`auth.claims` could originate from a JSON Web Token (JWT).
+
+### Parse and Check
+
+The parsing phase indicates whether the expression is syntactically valid and
+expands any macros present within the environment. Parsing and checking is
+more computationally expensive than evaluation, and it is recommended that
+expressions be parsed and checked ahead of time.
+
+```go
+parsed, issues := env.Parse(
+    `resource.name.startsWith("/groups/" + auth.claims.group)`)
+if issues != nil && issues.Err() != nil {
+    log.Fatalf("parse error: %s", issues.Err())
+}
+checked, issues := env.Check(parsed)
+if issues != nil && issues.Err() != nil {
+    log.Fatalf("type-check error: %s", issues.Err())
+}
+prg, err := env.Program(checked)
+if err != nil {
+    log.Fatalf("program construction error: %s", err)
 }
 ```
 
-More examples like these can be found within the unit tests which can be run
-using [Bazel][5]:
+The `cel.Program` generated at the end of parse and check is stateless,
+thread-safe, and cachable.
+
+Type-checking in an optional, but strongly encouraged step that can reject some
+semantically invalid expressions using static analysis. Additionally, the check
+produces metadata which can improve function invocation performance and object
+field selection at evaluation-time.
+
+#### Macros
+
+Macros are enabled by default and may be disabled. The comprehension macros
+`all`, `exists`, `exists_one`, `filter`, and `map` are particularly useful for evaluating a single predicate against list and map values.
+
+### Evaluate
 
+Now, evaluate for fun and profit. The evaluation is thread-safe and side-effect
+free. Many different inputs can be send to the same `cel.Program` and if fields
+are present in the input, but not referenced in the expression, they are
+ignored.
+
+```go
+// The `out` var contains the output of a successful evaluation.
+// The `details' var would contain intermediate evalaution state if enabled as
+// a cel.ProgramOption. This can be useful for visualizing how the `out` value
+// was arrive at.
+out, details, err := prg.Eval(map[string]interface{}{
+    "resource.name": "/groups/acme.co/documents/secret-stuff",
+    "auth.claims": map[string]interface{}{
+        "sub": "alice@acme.co",
+        "group": "acme.co"}})
+fmt.Println(out) // 'true'
 ```
+
+For more examples of how to use CEL, see [cel_test.go](https://github.com/google/cel-go/tree/master/cel/cel_test.go).
+
+#### Partial State
+
+What if the `auth.claims` hadn't been supplied? Well, CEL is designed for this.
+When some of the input is unknown, the `out` value will contain a list of the
```

Comment:
...when some input is bound to a special "Unknown" value.  Current behavior is a bug and shouldn't be documented.